### ISSUE TITLE
`isinstance` narrowing for `Generic` `NamedTuple` fails when narrowing against other tuple

### ISSUE THREAD

**Issue by colehaus** [microsoft/pyright#5346]
Reproduction code is as follows:

```py
from dataclasses import dataclass
from typing import Generic, NamedTuple, Optional, TypeVar

X = TypeVar("X")

class Foo(NamedTuple, Generic[X]):
    pass

@dataclass(frozen=True)
class Baz(Generic[X]):
    pass

def bar(foo: Foo[str] | tuple[int, int]) -> Optional[Foo[str]]:
    if isinstance(foo, Foo):
        return foo # pyright 1.1.314 complains that 'error: Return type, "Foo[str] | Foo[Unknown]", is partially unknown (reportUnknownVariableType)'                           
    else:
        return None

def corge(foo: Foo[str] | int) -> Optional[Foo[str]]:
    if isinstance(foo, Foo):
        return foo
    else:
        return None

def qux(baz: Baz[str] | tuple[int, int]) -> Optional[Baz[str]]:
    if isinstance(baz, Baz):
        return baz
    else:
        return None
```

The `isinstance` in `bar` doesn't seem to narrow as expected (see the error message in the inline comment)—it "forgets" that the only valid `Foo` for `foo` is `Foo[str]` given the type declaration for the function. Similar code in `corge` (other type in union isn't a tuple) and `qux` (`dataclass` instead of `NamedTuple`) does type check. `isinstance` does distinguish between the `NamedTuple` and the anonymous tuple at runtime.

mypy 1.3.0 successfully type checks all of these.

**Comment by erictraut** [comment]
I think pyright is correct here. Consider the following:

```python
from typing import Generic, NamedTuple, TypeVar

X = TypeVar("X")

class Foo(NamedTuple, Generic[X]):
    a: X
    b: X

def test(foo: Foo[str] | tuple[int, int]):
    if isinstance(foo, Foo):
        print("Is a Foo!")
    else:
        print("Is not a Foo!")

f = Foo[int](1, 2)
test(f) # Prints "Is a Foo!", but it's not a Foo[str]
```

The `NamedTuple` type is a `tuple` at runtime, so a named tuple is a subtype of `tuple`.

**Comment by erictraut** [comment]
Thinking about this more... There's a critical difference between your code sample and the code sample I posted above. I added two entries to the named tuple `Foo`, but in your example, `Foo` was defined as a zero-entry tuple. That means `Foo` in your example cannot ever be type compatible with `tuple[int, int]`. So I think it's safe to filter this subtype from the union when narrowing.

I've updated pyright's narrowing logic accordingly. It now produces the results you are looking for, and it still correctly handles the code sample that I posted.

This change will be included in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.315, which I just published. It will also be included in this week's insiders release of pylance.
